postgis/lwgeom_window.c

```diff
@@ -451,3 +451,423 @@ Datum ST_ClusterKMeans(PG_FUNCTION_ARGS)
+		}
+		else if (mode == COVERAGE_ISVALID)
+		{
+			GEOSCoverageIsValid(input, tolerance, &output);
+		}
+		else
+		{
+			elog(ERROR, "Unknown mode, never get here");
+		}
+
+		GEOSGeom_destroy(input);
+
+		if (!output)
+		{
+			HANDLE_GEOS_ERROR("Failed to process collection");
+		}
+
+		oldcontext = MemoryContextSwitchTo(uppercontext);
+		context->geom = (LWCOLLECTION *) GEOS2LWGEOM(output, GEOSHasZ(output));
+		MemoryContextSwitchTo(oldcontext);
+		GEOSGeom_destroy(output);
+
+		context->isdone = true;
+	}
+
+	/* Bomb out of any errors */
+	if (context->isnull)
+		PG_RETURN_NULL();
+
+	/* Propogate the null entries */
+	if (context->idx[curpos] < 0)
+		PG_RETURN_NULL();
+
+	/*
+	 * Geometry serialization is not const-safe! (we
+	 * calculate bounding boxes on demand) so we need
+	 * to make sure we have a persistent context when
+	 * we call the serialization, lest we create dangling
+	 * pointers in the object. It's possible we could
+	 * ignore them, skip the manual lwcollection_free
+	 * and let the aggcontext deletion take
+	 * care of the memory.
+	 */
+	oldcontext = MemoryContextSwitchTo(uppercontext);
+	subgeom = lwcollection_getsubgeom(context->geom, context->idx[curpos]);
+	if (mode == COVERAGE_ISVALID && lwgeom_is_empty(subgeom))
+	{
+		result = NULL;
+	}
+	else
+	{
+		result = geometry_serialize((LWGEOM*)subgeom);
+	}
+	MemoryContextSwitchTo(oldcontext);
+
+	/* When at the end of the partition, release the */
+	/* simplified collection we have been reading. */
+	if (curpos == rowcount - 1)
+		lwcollection_free(context->geom);
+
+	if (!result) PG_RETURN_NULL();
+
+	PG_RETURN_POINTER(result);
+
+}
+#endif /* POSTGIS_GEOS_VERSION >= 31200 */
+
+
+extern Datum ST_CoverageSimplify(PG_FUNCTION_ARGS);
+PG_FUNCTION_INFO_V1(ST_CoverageSimplify);
+Datum ST_CoverageSimplify(PG_FUNCTION_ARGS)
+{
+#if POSTGIS_GEOS_VERSION < 31200
+
+	lwpgerror("The GEOS version this PostGIS binary "
+		"was compiled against (%d) doesn't support "
+		"'GEOSCoverageSimplifyVW' function (3.12.0+ required)",
+		POSTGIS_GEOS_VERSION);
+	PG_RETURN_NULL();
+
+#else /* POSTGIS_GEOS_VERSION >= 31200 */
+
+	return coverage_window_calculation(fcinfo, COVERAGE_SIMPLIFY);
+
+#endif
+}
+
+
+extern Datum ST_CoverageInvalidLocations(PG_FUNCTION_ARGS);
+PG_FUNCTION_INFO_V1(ST_CoverageInvalidLocations);
+Datum ST_CoverageInvalidLocations(PG_FUNCTION_ARGS)
+{
+#if POSTGIS_GEOS_VERSION < 31200
+
+	lwpgerror("The GEOS version this PostGIS binary "
+		"was compiled against (%d) doesn't support "
+		"'GEOSCoverageIsValid' function (3.12.0+ required)",
+		POSTGIS_GEOS_VERSION);
+	PG_RETURN_NULL();
+
+#else /* POSTGIS_GEOS_VERSION >= 31200 */
+
+	return coverage_window_calculation(fcinfo, COVERAGE_ISVALID);
+
+#endif
+}
+
+/**********************************************************************
+ * ST_CoverageUnion(geometry[])
+ *
+ */
+
+Datum ST_CoverageUnion(PG_FUNCTION_ARGS);
+PG_FUNCTION_INFO_V1(ST_CoverageUnion);
+Datum ST_CoverageUnion(PG_FUNCTION_ARGS)
```

Comment:
Would this make more sense in `lwgeom_geos.c` ?

Reply:
I dunno that the extra layer of indirection is a big win? What's the argument for?

Reply:
Potato / potato. Could move it but then it would be separate from its coverage friends...